pkg/experiment/metastore/admin/admin.go

```diff
@@ -0,0 +1,258 @@
+
+	numRaftNodes := 0
+	nodes := make([]*metastoreNode, 0, len(a.servers))
+
+	for i, s := range a.servers {
+		cl := a.clients[i]
+		if cl == nil {
+			level.Warn(a.logger).Log("msg", "missing client for server", "server", s)
+			continue
+		}
+		node := &metastoreNode{
+			DiscoveryServerId: string(s.Raft.ID),
+			ResolvedAddress:   s.ResolvedAddress,
+		}
+		nodes = append(nodes, node)
+
+		res, err := cl.NodeInfo(ctx, &raftnodepb.NodeInfoRequest{})
+		if err != nil {
+			level.Warn(a.logger).Log("msg", "error fetching node info", "server", s, "err", err)
+			continue
+		}
+		nInfo := res.Node
+
+		node.RaftServerId = nInfo.ServerId
+		node.Member = nInfo.LeaderId != ""
+		node.State = nInfo.State
+		node.ObservedLeader = nInfo.LeaderId
+		node.CurrentTerm = nInfo.CurrentTerm
+		node.LastIndex = nInfo.LastIndex
+		node.CommitIndex = nInfo.CommitIndex
+		node.AppliedIndex = nInfo.AppliedIndex
+		node.BuildVersion = nInfo.BuildVersion
+		node.BuildRevision = nInfo.BuildRevision
+		node.Stats = make(map[string]string)
+		for i, n := range nInfo.Stats.Name {
+			node.Stats[n] = nInfo.Stats.Value[i]
+		}
+
+		if node.Member {
+			numRaftNodes++
+		}
+
+		// Each node has its own view of the world and can be out of sync with the rest.
+		// We choose the leader to be the one that is observed as the leader the most times.
+		leaderCounts[node.ObservedLeader]++
+		if leaderCounts[node.ObservedLeader] > maxLeaderCount && node.ObservedLeader != "" {
+			maxLeaderCount = leaderCounts[node.ObservedLeader]
+			leaderId = node.ObservedLeader
+		}
+	}
+	return &raftNodeState{
+		Nodes:    nodes,
+		LeaderId: leaderId,
+		NumNodes: numRaftNodes,
+	}, nil
+}
+
+func (a *MetastoreAdmin) newClient(address string) (raftnodepb.RaftNodeOpsServiceClient, error) {
+	// TODO aleks-p: do we need more configuration here?
+	client, err := grpc.NewClient(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+	return raftnodepb.NewRaftNodeOpsServiceClient(client), nil
+}
+
+func (a *MetastoreAdmin) addFormActionHandlers() {
+	a.actionHandlers["add"] = func(w http.ResponseWriter, r *http.Request, serverId string) error {
+		_, err := a.sharedClient.AddNode(r.Context(), &raftnodepb.AddNodeRequest{ServerId: serverId})
+		return err
+	}
+	a.actionHandlers["remove"] = func(w http.ResponseWriter, r *http.Request, serverId string) error {
+		raftState, err := a.fetchRaftState(r.Context())
+		if err != nil {
+			return err
+		}
+		if raftState.LeaderId == serverId {
+			// In theory, this shouldn't be needed since Raft should elect a new leader upon removal.
+			// In practice, a successful leader election doesn't always occur and demoting first is safer.
+			_, err = a.sharedClient.DemoteLeader(r.Context(), &raftnodepb.DemoteLeaderRequest{ServerId: serverId})
+			if err != nil {
+				return err
+			}
+		}
+		_, err = a.sharedClient.RemoveNode(r.Context(), &raftnodepb.RemoveNodeRequest{ServerId: serverId})
+		return err
+	}
```

Comment:
> We cannot do both raft.TransferLeadership and raft.RemoveServer in (*Node) RemoveNode because the second operation will fail once we lose leadership.

Sorry, my comment might be ambiguous. Please refer to my comment about the checks above.

In general, I believe we should avoid any logic in form handlers – their concern is the data representation.